### ISSUE TITLE
Postgres dialect does not recognise references to joined tables used in `json_object` function

### ISSUE THREAD

**Issue by PressXtoChris** [sqlfluff/sqlfluff#6656]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SqlFluff's postgres dialect fails to recognise references to joined tables used in the `json_object` function when using the syntax introduced in Postgres 16

### Expected Behaviour

SqlFluff should report no issues with this statement

### Observed Behaviour

`Joined table 'second_table AS st' not referenced elsewhere in querysqlfluff(ST11: structure.unused_join)`

### How to reproduce

```sql
SELECT json_object(
        'first': ft.str,
        'second': st.str
    ) AS json_column
FROM first_table AS ft
    LEFT JOIN second_table AS st
        ON ft.id = st.id;
```

### Dialect

postgres

### Version

3.3.1

### Configuration

```
[sqlfluff]
dialect = postgres
max_line_length = 120
exclude_rules = L039, L003, L071
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by joaostorrer** [comment]
I tested your SQL with sqlfluff 3.3.1 and ST11 rule is not being raised.

**Comment by PressXtoChris** [comment]
Strange, with version 3.3.1 I'm getting the error in both the CLI and the VSCode extension:

![Image](https://github.com/user-attachments/assets/a0da4e12-2f9f-4f50-9541-096def5667a0)

**Comment by keraion** [comment]
Fixed by #6708 which should be in 3.4.0
